bundles/org.openhab.binding.hive/src/main/java/org/openhab/binding/hive/internal/client/DefaultHiveClient.java

```diff
@@ -0,0 +1,160 @@
+
+        authenticate();
+    }
+
+    @Override
+    public void close() {
+        // TODO: Clean up
+    }
+
+    /**
+     * Try to authenticate with the Hive API.
+     *
+     * @throws org.openhab.binding.hive.internal.client.exception.HiveApiAuthenticationException
+     *      If we failed to authenticate with the stored username and password.
+     *
+     * @throws org.openhab.binding.hive.internal.client.exception.HiveApiUnknownException
+     *      If something unexpected happens while communicating with the Hive
+     *      API.
+     */
+    private void authenticate() {
+        this.authenticationManager.clearSession();
+
+        final Session session = this.sessionRepository.createSession(
+                this.username,
+                this.password
+        );
+
+        this.session = session;
+        this.authenticationManager.setSession(session);
+    }
+
+    private <T> T makeAuthenticatedApiCall(final Supplier<T> apiCall) {
+        // If we get a valid result return it.
+        // If we are not authorised check if session has expired, reauthenticate and try again.
+        // Otherwise let other exceptions bubble up.
+        boolean reauthenticated = false;
+        while (true) {
+            try {
+                return apiCall.get();
+            } catch (final HiveApiNotAuthorisedException ex) {
+                if (reauthenticated || this.sessionRepository.isValidSession(this.authenticationManager.getSession())) {
+                    // We are either not authorised for this resource or
+                    // something has gone wrong with the client logic.
+                    // Pass on the exception.
+                    throw ex;
+                } else {
+                    // Session seems to no longer be valid.
+                    // Reauthenticate and try again.
+                    authenticate();
+                    reauthenticated = true;
+                }
+            }
+        }
+    }
+
+    @Override
+    public UserId getUserId() {
+        final @Nullable Session session = this.session;
+        if (session == null) {
+            throw new IllegalStateException("Session is unexpectedly null.");
+        }
+
+        return session.getUserId();
+    }
+
+    @Override
+    public Set<Node> getAllNodes() {
+        return makeAuthenticatedApiCall(this.nodeRepository::getAllNodes);
+    }
+
+    @Override
+    public String getAllNodesJson() {
+        return makeAuthenticatedApiCall(this.nodeRepository::getAllNodesJson);
+    }
+
+    @Override
+    public @Nullable Node getNode(final NodeId nodeId) {
+        Objects.requireNonNull(nodeId);
+
+        // N.B. Type parameter because Checker Framework needs a little help.
+        return this.<@Nullable Node>makeAuthenticatedApiCall(() -> this.nodeRepository.getNode(nodeId));
+    }
+
+    @Override
+    public @Nullable Node updateNode(final Node node) {
+        Objects.requireNonNull(node);
```

Comment:
I ended up doing this out of habit. I don't really trust the ecj null checker because I have noticed times when it is very wrong (not warning about things things that can definitely be null).
Interestingly I found that Checker Framework integrates quite easily with this project and seems more rigorous (and comes with an annotated JDK). That's a discussion for another time though as you probably don't want individual bindings overriding the compiler configuration.